doc/decisions/0_drafts/keyname.md

```diff
@@ -0,0 +1,136 @@
+# Namespace and Name of Keys
+
+## Problem
+
+A `Key` in Elektra is identified by its name, which consists of a namespace and a number of name parts.
+There are two common representations for the name: escaped and unescaped.
+
+The unescaped form is essentially a single namespace byte plus and arbitrary-length sequence of arbitrary bytes, in which a zero byte separates parts.
+The escaped name is a more complicated representation as a zero-terminated string that maps 1:1 onto unescaped names.
+More details can be found in [the relevant docs](/doc/KEYNAMES.md).
+
+There is a conflict between these two forms in terms of API convenience and efficient execution.
+Generally, the escaped form is more convenient to use, since it is entirely human-readable and just a "normal" string.
+When it comes to execution, however, the unescaped name often makes things more efficient.
+For example, order comparisons between unescaped names are much simpler.
+
+The question now is, which representations should be used by `libelektra-core` and how.
+
+## Constraints
+
+- Because `KeySet` is ordered by name and stores `Key`, the order comparison between the name of two `Key`s must be fast.
+
+## Assumptions
+
+- In most cases the escaped name is used for developer convenience not, because of actual requirements.
+- The most common requirement for using the escaped name is UI: reading names from or displaying them in a user interface (e.g., `kdb` CLI)
+
+## Considered Alternatives
+
+### Only escaped name
+
+Because the escaped name is a simple zero-terminated string, it can be represented as a single `char *`.
+
+Storing the name as a single `char *` would be the most space efficient.
+But resizing would require counting the length every time.
+Therefore, for storage the better solution may be a `char *` and a `size_t`.
+
+However, in the API the name could always be a single `char *`, making for a very easy to use API.
+
+The biggest problem with this approach is that comparing two escaped names is not trivial.
+The comparison needs to account for namespaces, parts and escaping.
+
+Similarly, iterating over the individual parts of a name (and/or manipulating them) is non-trivial, because it requires logic to handle escape sequences.
+
+### Only unescaped name
+
+The unescaped name contains zero bytes.
```

Comment:
```suggestion
The unescaped name contains zero-bytes (`\0`).
```

Took me an embarrassingly long time to make sense of this sentence 😅

Reply:
Ah yes, that was badly phrased :sweat_smile: